File: tokenizers/cpp/bpe.cpp
#include <headers/bpe.hpp>
#include <stringUtil.hpp>
#include <json.hpp>

#include <map>
#include <vector>
#include <string>
#include <queue>
#include <iostream>
#include <algorithm>
#include <tuple>
#include <fstream>

using json = nlohmann::json;

Word::Word(DoubleLinkedList dll, int freq) {
    this->dll = dll;
    this->freq = freq;
}




BPE::BPE() {
    // does nothing
}


BPE::~BPE() = default;


std::priority_queue<std::pair<int, std::pair<std::string, std::string>>> BPE::create_pair() {
    for(auto& element : BPE::words) {
        Node* curr = element.dll.head;

        while(curr->data != "</w>") {
            std::pair<std::string, std::string> curr_pair = {curr->data, curr->next->data};
            BPE::pair_to_word[curr_pair].insert(element);
            BPE::pair_freq[curr_pair] += element.freq;
            curr = curr->next;
        }
    }

    std::priority_queue<std::pair<int, std::pair<std::string, std::string>>> pq;
    for(auto& it : BPE::pair_freq) {
        pq.push({it.second, it.first});
    }

    return pq;
}


void BPE::merge(std::pair<int, std::pair<std::string, std::string>>& candidate, int rank) {
    auto pair = candidate.second;
    BPE::merge_rule[pair] = rank;

    std::vector<Word> local_copy(BPE::pair_to_word[pair].begin(), BPE::pair_to_word[pair].end());
    for(auto& element : local_copy) {
        Node* curr = element.dll.head;

        while(curr != nullptr && curr->next != nullptr) {
            if(curr->data == pair.first && curr->next->data == pair.second) {
                BPE::pair_freq[pair] -= element.freq;
                BPE::pq.push({BPE::pair_freq[pair], pair});
                
                curr->data = pair.first + pair.second;

                if(curr->next->next != nullptr) {
                    BPE::pair_freq[{pair.second, curr->next->next->data}] -= element.freq;
                    BPE::pq.push({BPE::pair_freq[{pair.second, curr->next->next->data}], {pair.second, curr->next->next->data}});

                    curr->next = curr->next->next;
                    curr->next->prev = curr;

                    BPE::pair_freq[{curr->data, curr->next->data}] += element.freq;
                    BPE::pair_to_word[{curr->data, curr->next->data}].insert(element);
                    BPE::pq.push({BPE::pair_freq[{curr->data, curr->next->data}], {curr->data, curr->next->data}});
                }
                else {
                    curr->next = nullptr;
                    element.dll.tail = curr;
                }

                if(curr->prev != nullptr) {
                    BPE::pair_freq[{curr->prev->data, pair.first}] -= element.freq;
                    BPE::pq.push({BPE::pair_freq[{curr->prev->data, pair.first}], {curr->prev->data, pair.first}});

                    BPE::pair_freq[{curr->prev->data, curr->data}] += element.freq;
                    BPE::pair_to_word[{curr->prev->data, curr->data}].insert(element);
                    BPE::pq.push({BPE::pair_freq[{curr->prev->data, curr->data}], {curr->prev->data, curr->data}});
                }
            }

            curr = curr->next;
        }
    }

    BPE::pair_freq.erase(pair);
    BPE::pair_to_word.erase(pair);

    return;
}


void BPE::train(std::vector<std::string>& corpus, int num_merges) {
    std::map<std::string, int> word_freq;
    for(auto& it : corpus) {
        word_freq[it] ++;
    }

    for(auto& it : word_freq) {
        DoubleLinkedList dll;

        for(auto& ch : it.first) {
            dll.push_back(std::to_string(ch));
        }
        dll.push_back("</w>");

        Word element(dll, it.second);
        BPE::words.push_back(element);
    }

    BPE::pq = BPE::create_pair();

    int rank = 0;
    while(!BPE::pq.empty() && num_merges > 0) {
        auto candidate = BPE::pq.top();
        BPE::pq.pop();

        if(candidate.first > 0 && candidate.first == BPE::pair_freq.count(candidate.second)) {
            rank += 1;
            BPE::merge(candidate, rank);
            num_merges -= 1;
        }
    }

    std::cout << "BPE Done" << std::endl;

    return;
}


std::vector<std::string> BPE::encode(std::string& text) {
    std::vector<std::string> words_1 = StringUtil::split(text, ' ');
    std::vector<DoubleLinkedList> words_2(words_1.size());

    for(size_t i = 0; i < words_1.size(); i ++) {
        DoubleLinkedList dll;

        for(auto& ch : words_1[i]) {
            dll.push_back(std::string(1, ch));
        }
        dll.push_back("</w>");

        words_2[i] = dll;
    }

    std::vector<std::string> encoded_text;

    for(auto& word : words_2) {
        std::priority_queue<std::tuple<int, int, Node*, std::string, std::string>,
                            std::vector<std::tuple<int, int, Node*, std::string, std::string>>,
                            std::greater<std::tuple<int, int, Node*, std::string, std::string>>> pq;

        Node* curr = word.head;
        int tiebreak = 0;

        while(curr->data != "</w>") {
            int rank = INT_MAX;
            if(auto it = BPE::merge_rule.find({curr->data, curr->next->data}); it != BPE::merge_rule.end()) {
                rank = it->second;
            }

            if(rank != INT_MAX) {
                tiebreak += 1;
                pq.push({rank, tiebreak, curr, curr->data, curr->next->data});
            }

            curr = curr->next;
        }

        while(!pq.empty()) {
            auto smallest = pq.top();
            pq.pop();

            Node* node = std::get<2>(smallest);

            if(node->next != nullptr && node->data == std::get<3>(smallest) && node->next->data == std::get<4>(smallest)) {
                node->data = std::get<3>(smallest) + std::get<4>(smallest);

                if(node->next->next != nullptr) {
                    node->next = node->next->next;
                    node->next->prev = node;
                }
                else {
                    node->next = nullptr;
                    word.tail = node;
                }

                if(node->prev != nullptr) {
                    int rank = INT_MAX;
                    if(auto it = BPE::merge_rule.find({node->prev->data, node->data}); it != BPE::merge_rule.end()) {
                        rank = it->second;
                    }

                    if(rank != INT_MAX) {
                        tiebreak += 1;
                        pq.push({rank, tiebreak, node->prev, node->prev->data, node->data});
                    }
                }
            }
        }

        curr = word.head;
        while(curr != nullptr) {
            encoded_text.push_back(curr->data);
            curr = curr->next;
        }
    }

    return encoded_text;
}


std::string BPE::decode(std::vector<std::string>& encoded_text) {
    std::string full_text = "";
    for (const auto& token : encoded_text) {
        full_text += token;
    }

    std::vector<std::string> words = StringUtil::split(full_text, "</w>");

    std::string decoded_output = "";
    size_t limit = words.empty() ? 0 : words.size() - 1;

    for (size_t i = 0; i < limit; i++) {
        if (i > 0) {
            decoded_output += " ";
        }
        decoded_output += words[i];
    }

    return decoded_output;
}


void BPE::export_merges(std::string path) {
    std::vector<std::pair<std::pair<std::string, std::string>, int>> sorted_rules(BPE::merge_rule.begin(), BPE::merge_rule.end());

    std::sort(sorted_rules.begin(), sorted_rules.end(), 
        [](const auto& a, const auto& b) {
            return a.second < b.second; 
        }
    );

    json j_pairs = json::array();

    for (const auto& rule : sorted_rules) {
        json pair_arr = {rule.first.first, rule.first.second};
        j_pairs.push_back(pair_arr);
    }

    std::ofstream file(path);

    if (file.is_open()) {
        file << j_pairs.dump(2); 
        file.close();
        std::cout << "Merge rules are saved at : " << path << std::endl;
    }
    else {
        std::cerr << "Error: Could not open file for writing at " << path << std::endl;
    }

    return;
}

void BPE::load_merges(std::string path) {
    std::ifstream file(path);

    if (!file.is_open()) {
        std::cerr << "Error: Could not open file for reading at " << path << std::endl;
        
        return;
    }

    json j_pairs;
    file >> j_pairs;
    file.close();

    BPE::merge_rule.clear();

    int rank = 0;
    for (const auto& item : j_pairs) {
        
        if (item.is_array() && item.size() == 2) {
            std::string a = item[0].get<std::string>();
            std::string b = item[1].get<std::string>();
            
            BPE::merge_rule[{a, b}] = rank;
            rank++;
        }
    }

    std::cout << "Merge rules loaded from: " << path << std::endl;

    return;
}

File: tokenizers/bpe.py
from collections import Counter, defaultdict
import math
from utilities.DoubleLinkedList import DoubleLinkedList
import heapq
import json
import warnings


class Word:
    def __init__(self, dll, freq):
        self.dll = dll
        self.freq = freq




class BPE():
    def __init__(self):
        self.merge_rule = {} # map to rank the merges
        self.pair_freq = defaultdict(int) # map : tuple -> count
        self.pair_to_word = defaultdict(set) # map : tuple -> set
        self.words = []

    
    def __create_pair(self):
        for element in self.words:
            curr = element.dll.head

            while curr.data != '</w>':
                pair = (curr.data, curr.next.data)
                self.pair_to_word[pair].add(element)
                self.pair_freq[pair] += element.freq
                curr = curr.next
            
        pq = [] # max heap : (-frequence, pair)
        for pair in self.pair_freq.keys():
            heapq.heappush(pq, (-self.pair_freq[pair], pair))
        
        return pq


    def __merge(self, candidate, rank):
        pair = candidate[1]
        self.merge_rule[pair] = rank

        for element in set(self.pair_to_word[pair]):
            curr = element.dll.head

            while curr is not None and curr.next is not None:

                if curr.data == pair[0] and curr.next.data == pair[1]:
                    self.pair_freq[pair] -= element.freq
                    heapq.heappush(self.pq, (-self.pair_freq[pair], pair))

                    curr.data = pair[0] + pair[1]

                    if curr.next.next is not None:
                        self.pair_freq[(pair[1], curr.next.next.data)] -= element.freq
                        heapq.heappush(self.pq, (-self.pair_freq[(pair[1], curr.next.next.data)], (pair[1], curr.next.next.data)))

                        curr.next = curr.next.next
                        curr.next.prev = curr

                        self.pair_freq[(curr.data, curr.next.data)] += element.freq
                        self.pair_to_word[(curr.data, curr.next.data)].add(element)
                        heapq.heappush(self.pq, (-self.pair_freq[(curr.data, curr.next.data)], (curr.data, curr.next.data)))
                    else:
                        curr.next = None
                        element.dll.tail = curr
                    
                    if curr.prev is not None:
                        self.pair_freq[(curr.prev.data, pair[0])] -= element.freq
                        heapq.heappush(self.pq, (-self.pair_freq[(curr.prev.data, pair[0])], (curr.prev.data, pair[0])))

                        self.pair_freq[(curr.prev.data, curr.data)] += element.freq
                        self.pair_to_word[(curr.prev.data, curr.data)].add(element)
                        heapq.heappush(self.pq, (-self.pair_freq[(curr.prev.data, curr.data)], (curr.prev.data, curr.data)))

                curr = curr.next
        
        self.pair_freq.pop(pair, None)
        self.pair_to_word.pop(pair, None)

        return


    def train(self, corpus, num_merges):
        word_freq = dict(Counter(corpus))
        
        for word in word_freq.keys():
            dll = DoubleLinkedList()

            for character in list(word) + ['</w>']:
                dll.push_back(character)

            element = Word(dll, word_freq[word])
            self.words.append(element)

        self.pq = self.__create_pair()

        rank = 0
        while self.pq and num_merges > 0:
            candidate = heapq.heappop(self.pq)

            ## validate before merging
            if -candidate[0] > 0 and -candidate[0] == self.pair_freq.get(candidate[1], 0):
                rank += 1
                self.__merge(candidate, rank)
                num_merges -= 1
        
        print("BPE Done!")

        return


    def encode(self, text):
        words = text.split(" ")

        for i in range(len(words)):
            dll = DoubleLinkedList()

            for character in list(words[i]) + ['</w>']:
                dll.push_back(character)

            words[i] = dll

        encoded_text = []

        for word in words:
            pq = []
            curr = word.head
            tiebreak = 0

            while curr.data != '</w>':
                rank = self.merge_rule.get((curr.data, curr.next.data), math.inf)

                if rank != math.inf:
                    tiebreak += 1
                    heapq.heappush(pq, (rank, tiebreak, curr, curr.data, curr.next.data))
                curr = curr.next
            
            while pq:
                smallest = heapq.heappop(pq)
                node =  smallest[2]

                ## validate before merging
                if node.next is not None and node.data == smallest[3] and node.next.data == smallest[4]:
                    node.data = smallest[3] + smallest[4]

                    ## validate if we have merged the last two pairs or not
                    if node.next.next is not None:
                        node.next = node.next.next
                        node.next.prev = node
                    else:
                        node.next = None
                        word.tail = node

                    ## push the new pairs
                    ## check if the next element exists or not
                    if node.next is not None:
                        rank = self.merge_rule.get((node.data, node.next.data), math.inf)

                        if rank != math.inf:
                            tiebreak += 1
                            heapq.heappush(pq, (rank, tiebreak, node, node.data, node.next.data))

                    ## check if the previous element exists or not
                    if node.prev is not None:
                        rank = self.merge_rule.get((node.prev.data, node.data), math.inf)

                        if rank != math.inf:
                            tiebreak += 1
                            heapq.heappush(pq, (rank, tiebreak, node.prev, node.prev.data, node.data))

            curr = word.head
            
            while curr is not None:
                encoded_text.append(curr.data)
                curr = curr.next

        return encoded_text


    def decode(self, encoded_text):
        dec_str_1 = ["".join(word) for word in encoded_text]
        dec_str_2 = "".join(dec_str_1)
        
        output = dec_str_2.split('</w>')[ : -1]
        decoded_output = " ".join(output)

        return decoded_output
    
    def export_merges(self, path):
        # merge_rule is {(a,b): rank}; emit pairs ordered by rank
        pairs_in_order = [list(pair) for pair, _ in sorted(self.merge_rule.items(), key=lambda kv: kv[1])]
        with open(path, "w") as f:
            json.dump(pairs_in_order, f, indent=2)
        
        print(f"Merge rules are saved at : {path}")

        return

    def load_merges(self, path):
        with open(path) as f:
            pairs = json.load(f)
        self.merge_rule = {(a, b): rank for rank, (a, b) in enumerate(pairs)}

        print(f"Merge rules loaded from: {path}")

        return


if __name__ == "__main__":

    BPE_tokenizer = BPE()

    BPE_tokenizer.train([
        "low", "low", "low", "low", "low",
        "lower", "lower",
        "newest", "newest", "newest", "newest", "newest", "newest",
        "widest", "widest", "widest"], 10)
    
    BPE_tokenizer.export_merges("tokenizers/merge_rules.json")
    BPE_tokenizer.load_merges("tokenizers/merge_rules.json")

    test_text = "low newest zero"
    encoded_text = BPE_tokenizer.encode(test_text)
    decoded_text = BPE_tokenizer.decode(encoded_text)
    print(f"Original text: {test_text}")
    print(f"Encoded text: {encoded_text}")
    print(f"Decoded text: {decoded_text}")